Abscal run rewrite

jsdillon opened the pull request:
* Rewrites and simplifies abscal run functions to only do post-redcal abscal.
* Enables partial I/O in abscal, partially addressing #319.
* Better normalizes abscal chi^2 relative to the noise generated from autocorrelations. Does not properly take into account the degrees of freedom, which is non-trivial to count. I think this is good enough for now, so I say this closes #375.
* Merges abscal.py and abscal_funcs.py. This closes #355.
* With the new tests, I'd say this also close #262.

I think with this, we can finally close #335.

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+    '''
+    # first produce a map of indices using the LSTs
+    m2d_ind_map = {}  
+    for dind, dlst in enumerate(data_lsts):
```
should consider the effects of LST wrap in this FOR loop

jsdillon replied:
Done.

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+            otherwise, 'multiply'.
+        max_iter: maximum number of times to run phase solvers iteratively to avoid the effect
+            of phase wraps in, e.g. phase_slope_cal or TT_phs_logcal
+        phs_conv_crit: convergence criterion for updates to iterative phase calibration that compares
```
function isn't specific to phs, so should be `conv_crit`

jsdillon replied:
Actually, the way I wrote it, it's very specific to phase. The non-phase steps don't suffer from phase wraps, which was the original motivation for making them iterative (unlike redcal, where the iterative steps address the fundamentally non-linear nature of redcal).

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+        if max_iter > 1:
+            crit = np.median(np.linalg.norm([gains_here[k] - 1.0 for 
+                                             k in gains_here.keys()], axis=(0, 1)))
+            echo("phase_slope_cal convergence criterion: " + str(crit), verbose=verbose)
```
put `AC_func` in here instead of `phase..`

jsdillon replied:
Good call. Done.

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+    AC.antpos = data.antpos
+
+    # Global Delay Slope Calibration
+    for time_avg in [True, False]:
```
haven't exposed `time_avg` kwarg, and why do you need the conditional?

jsdillon replied:
I thought the idea was to get an approximate global delay slope (assuming it doesn't change) by doing the average, then do it again for each integration. That's exactly what the old abscal run did.

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+        add_to_history: string to add to history of output abscal file
+
+    Returns:
+        hc: HERACal object which was written to disk. Matches the input redcal_file with an updated history and:
```
this makes it seem like 5 things are returned, perhaps consider another way of phrasing this w/o making it seem like `gains` `flags` etc are returned

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+        model_files: list of string paths to externally calibrated data. Strings must be sortable 
+            to produce a chronological list in LST (wrapping over 2*pi is OK)
+        output_replace: tuple of two strings to find and replace in redcal_file to produce the output calfits file
+        nInt_to_load: number of integrations to load and calibrate simultaneously. Default -1 loads all integrations.
```
can this be `None` instead of -1, I'm used to `-1` being an axis specification, and also `None` follows `io.HERAData` convention when specifying `bls` `pols` etc..

jsdillon replied:
Done. I also updated redcal to have a consistent convention.

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+    '''Perform abscal on entire data files, picking relevant model_files from a list and doing partial data loading.
+    
+    Arguments:
+        data_file: string path to raw uvh5 visibility file
```
why are we limiting ourselves to only `uvh5`? why not make this flexible? One use-case is we want to compare this to IDR2.1 outputs

jsdillon replied:
I'd like to punt in more data formats. See below.

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+    
+    Arguments:
+        data_file: string path to raw uvh5 visibility file
+        redcal_file: string path to calfits file that redundantly calibrates the data_file
```
this should be optional, in case the data are already calibrated

jsdillon replied:
I'd like to punt on this too. See https://github.com/HERA-Team/hera_cal/pull/430#issuecomment-457373546

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+        redcal_file: string path to calfits file that redundantly calibrates the data_file
+        model_files: list of string paths to externally calibrated data. Strings must be sortable 
+            to produce a chronological list in LST (wrapping over 2*pi is OK)
+        output_replace: tuple of two strings to find and replace in redcal_file to produce the output calfits file
```
let's just make this more flexible and easier and be `output_fname`

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+    if os.path.exists(redcal_file.replace(*output_replace)) and not clobber:
+        raise IOError("{} exists, not overwriting.".format(redcal_file.replace(*output_replace)))
+
+    # Load redcal calibration
```
if fed

jsdillon replied:
I'd like to punt on this too. See #430 (comment)

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
-
-    # make blank gains if no modelfiles
+def get_all_times_and_lsts(hd, solar_horizon=90.0, unwrap=True):
+    '''Extract all times and lsts from a HERAData object
```
this only works for a `HERAData` instantiated with a `uvh5`, see `io.get_file_lsts` for how to do this for other file formats. Let's just have one function in the codebase to do this functionality, if you want to add `times` in addition to `lsts`, great!

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+        phs_conv_crit: convergence criterion for updates to iterative phase calibration that compares them to all 1.0s.
+        clobber: if True, overwrites existing abscal calfits file at the output path
+        add_to_history: string to add to history of output abscal file
+
```
should provide `read_kwargs` to pass to `HERAData.read` calls on model and data. example is running only a single pol from multi-pol files

jsdillon replied:
Again, I think this is something to punt on and open as an issue if it ever becomes useful or necessary.

It's fairly trivial to rewrite your redcal and data files with a pyuvdata select before running this function. As it stands, if the number of polarizations to run on is less than what's in either data or redcal file, it's not obvious how it should behave. Again, more complexity than I think is likely to be useful.

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+    # Raise error if output calfile already exists and clobber is False
+    if os.path.exists(redcal_file.replace(*output_replace)) and not clobber:
+        raise IOError("{} exists, not overwriting.".format(redcal_file.replace(*output_replace)))
+
```
partial IO doesn't work for `miriad` files across time, so there should be a check on `nInt == None` if `filetype` == `miriad`

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+def post_redcal_abscal_run(data_file, redcal_file, model_files, output_replace=('.omni.', '.abs.'), nInt_to_load=-1,
+                           data_solar_horizon=90, model_solar_horizon=90, min_bl_cut=1.0, max_bl_cut=None, edge_cut=0, 
+                           tol=1.0, phs_max_iter=100, phs_conv_crit=1e-6, clobber=True, add_to_history='', verbose=True):
+    '''Perform abscal on entire data files, picking relevant model_files from a list and doing partial data loading.
```
might want to mention that this doesn't work for BDA files

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+    # match times to narrow down model_files
+    matched_model_files = sorted(set(match_times(data_file, model_files, filetype='uvh5')))
+    if len(matched_model_files) > 0:
+        hd = io.HERAData(data_file)
```
This gets a little tricky for non `uvh5` files b/c `pols` won't be loaded by default. We could do a hack and load a single integration, which would be quick and also load the relevant metadata if `filetype==miriad`

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+    if len(matched_model_files) > 0:
+        hd = io.HERAData(data_file)
+        hdm = io.HERAData(matched_model_files)
+        pol_load_list = [pol for pol in hd.pols if split_pol(pol)[0] == split_pol(pol)[1]]
```
update given `polarizations` kwarg

jsdillon replied:
See https://github.com/HERA-Team/hera_cal/pull/430#discussion_r250795758

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+
+                        # apply calibration
+                        data_ants = set([ant for bl in data.keys() for ant in split_bl(bl)])
+                        rc_gains_subset = {k: rc_gains[k][tinds, :] for k in data_ants}
```
if fed `input_cal`

jsdillon replied:
See https://github.com/HERA-Team/hera_cal/pull/430#issuecomment-457373546

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+                                           cal_flags=rc_flags_subset, gain_convention=hc.gain_convention)
+                        
+                        # run absolute calibration, copying data because it gets modified internally
+                        delta_gains, AC = post_redcal_abscal(model, copy.deepcopy(data), flags, rc_flags_subset, edge_cut=edge_cut, 
```
lets think if this deepcopy is really needed. I don't think it is. Just use `AC.data` and get rid of second `calibrate_in_place`

jsdillon replied:
The issue here is that I need calibrated autocorrelations to normalize chi^2... which are often deleted from AC.data. However, I think I can solve the issue in a memory-efficient way by deepcopying the autos before creating the Abscal object, rather than deepcopying the whole dataset. Thanks for making me think it through.

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+                        # update results
+                        delta_flags = synthesize_ant_flags(flags)
+                        for ant in data_ants:
+                            abscal_gains[ant][tinds, :] = rc_gains_subset[ant] * delta_gains[ant]
```
I think you're double counting `rc_gains` here b/c `abscal_gains` was a deepcopy to begin with.

jsdillon replied:
No, because I'm doing = rather than *=, so I'm overwriting.

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+        phs_max_iter: maximum number of iterations of phase_slope_cal or TT_phs_cal allowed
+        phs_conv_crit: convergence criterion for updates to iterative phase calibration that compares
+            the updates to all 1.0s. 
+
```
let's add a `refant` kwargs that is `None` by default which triggers `pick_reference_antenna`

jsdillon replied:
I added refant.

I don't think exposing that kwarg is necessary. All I'm looking for is the least flagged antenna. If everything is flagged in synthesize_ant_flags somehow, it doesn't much matter.

nkern commented on hera_cal/io.py:
```diff
@@ -776,6 +776,34 @@ def get_file_lst_range(filepaths, filetype='uvh5', add_int_buffer=False):
     return file_starts, file_stops, int_times
 
 
+def partial_time_io(hd, times, **kwargs):
```
specify in docstring this must be a `uvh5`-instatiated `HERAData` object

nkern commented on hera_cal/io.py:
```diff
@@ -776,6 +776,34 @@ def get_file_lst_range(filepaths, filetype='uvh5', add_int_buffer=False):
+    combined_hd = None
+    for f in hd.filepaths:
+        hd_here = HERAData(f)
+        times_here = [t for t in times if t in hd_here.times]
```
can we have it catch the case of files that don't have partial io and thus `hd_here.times == None` and have it just load all times?

jsdillon replied:
It's not clear that that's what we want it to do. I'll add an assertion instead.

nkern commented on hera_cal/abscal.py:
```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
+                            abscal_chisq[antpol][tinds, :] = total_qual[antpol] / nObs[antpol]  # Note, not normalized for DoF
+                            
+        # impose a single reference antenna on the final antenna solution
+        refant = pick_reference_antenna(abscal_flags)
```
use `refant` if fed

jsdillon replied:
Done.

nkern commented:
A really great addition to `abscal`, making it partial IO capable across time, and significantly modularized the `abscal_run` drive function. Thanks!

I've posted some minor comments, and I have one major issue, which is that I think the `post_redcal_abscal` function is a *little* bit too specific for IDR2.2 run setup. I'm fine w/ making it specific in the sense of which `abscal` functions to run, hence `post_redcal_abscal` name, but even the inputs and outputs assumes the exact handoff of data products as we will do them in IDR2.2, and I think that's a little too rigid. I think we can make this better w/o making it more complicated by:

1) allowing more than just `uvh5` inputs, which takes a little bit of massaging for some of the sub-functions, and in some cases eliminates the partial io capability, which is fine b/c we aren't doing that anyways w/ `miriad` files

2) allowing for `input_cal` optionality, instead of rigidity.

3) allowing for `output_fname` instead of assuming that the output filename is an input calfits w/ a replaced keyword

If we were able to get those 3 things in, I would feel more comfortable w/ us erasing the old `abscal_run`, which is part of this PR.

Great work so far!

jsdillon commented:
Thanks so much @nkern for the quick review. I agree with most of your comments and I think I can implement most of this pretty straightforwardly. 

In the interest of time, I think I'm going to punt on non-uvh5 files. That's a bunch more work and in the interest of getting IDR2.2 done, I want to get this merged in ASAP. Also, redcal is uvh5-only. Converting between data formats is so easy with pyvudata, I imagine in most cases that's a better option than building in a lot more legacy structure. However, if you want to open an issue to update the post_redcal_abscal functions to support more file formats, go right ahead.

jsdillon commented:
Ok. I've addressed most of your comments. However, I've given it some more thought and it's not trivial to support pre-calibrated data. This code relies on the redcal_file to do a number of things:

1. Initializing empty dictionaries for holding the abscal results that are then 
2. Holding antenna-based flags used for calibrate_in_place
3. Specifying the gain convention.
4. Providing all the correct metadata for writing the results to disk with the appropriate history.

It's certainly possible to bypass all of that, but I don't think that extra flexibility is worth complicating the control structure of the code and incurring more overhead in terms of unit testing and maintainability. The more if statements, the more unit tests we need. 

I especially don't think it's worth holding up the PR for extra flexibility that we don't think we need right now and really shouldn't be tempted to use, since we're trying to save disk space by not re-rewriting calibrated data. After all, "There should be one--and preferably only one--obvious way to do it."
